File: 9.20/9.20block相关/9.20block相关/ViewController.m
//
//  ViewController.m
//  9.20block相关
//
//  Created by codygao on 16/9/21.
//  Copyright © 2016年 HM. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
@property(nonatomic,copy)void (^myblock)();

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self callbackWith:^{
        NSLog(@"定义回调函数");
    }];
  
    //定义block
    
    void (^myblock)() = ^{
        NSLog(@"全局的block到处跑");
    };
    self.myblock = myblock;
}
-(void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    [self blockdemo2];
    //调用全局的block
    self.myblock();
}
//当做参数来传递
-(void)blockdemo2{
    void (^task)() = ^{
        NSLog(@"当做参数在传递");
    };
    
}

//定义接收block的方法
-(void)callbackWith:(void (^)())task{
    NSLog(@"%s",__func__);
//    回调task
    task();
}

//简单的定义block

-(void)blockDemo1{
    
    int num = 10;
    void (^block)() = ^{
        NSLog(@"无参无返回值");
    };
    
    block();
    
    
    
    
}
@end

File: 9.17懒加载/9.17懒加载/ViewController.swift
//
//  ViewController.swift
//  9.17懒加载
//
//  Created by codygao on 16/9/18.
//  Copyright © 2016年 HM. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    //懒加载方式一
    var person: Person = Person()
    lazy var p: Person = {()->Person in
        let p1 = Person()
        p1.name = "我是懒加载出来的"
        return p1
    }()
    //懒加载方式二
    lazy var p1: Person = {
        let person = Person()
        person.name = "我也是懒加载出来的"
        return person
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
    
        print(p.name)
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        print(p1.name)
    }
    

}

